# Dataset Sonar

Some classification problem on sonar.csv dataset

In [1]:
import os
import pandas as pd
import matplotlib.pyplot as plt
import imblearn
from sklearn.preprocessing import StandardScaler, PowerTransformer
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC, LinearSVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB

In [2]:
directory = os.path.join('D:/Bootcamp ML - Mada/classification dataset/', 'sonar.csv')
if os.path.isfile(directory):
  print("File ditemukan")
else:
    print("tidak ada")

File ditemukan


In [3]:
dataset = pd.read_csv(directory)
dataset

V1      V2      V3      V4      V5      V6      V7      V8      V9  \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

        V10  ...     V52     V53     V54     V55     V56     V57     V58  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

        V59     V60  Class  
0    0.0090  0.0032      1  
1    0.0052  0.0044      1  
2    0.0095  0.0078      1  
3    0.0040  0.0117      1  
4    0.0107  0.0094      1  
..      ...     ...    ...  
203  0.0193  0.0157      0  
204  0.0062  0.0067      0  
205  0.0077  0.0031      0  
206  0.0036  0.0048      0  
207  0.0061  0.0115      0  

[208 rows x 61 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      208 non-null    float64
 1   V2      208 non-null    float64
 2   V3      208 non-null    float64
 3   V4      208 non-null    float64
 4   V5      208 non-null    float64
 5   V6      208 non-null    float64
 6   V7      208 non-null    float64
 7   V8      208 non-null    float64
 8   V9      208 non-null    float64
 9   V10     208 non-null    float64
 10  V11     208 non-null    float64
 11  V12     208 non-null    float64
 12  V13     208 non-null    float64
 13  V14     208 non-null    float64
 14  V15     208 non-null    float64
 15  V16     208 non-null    float64
 16  V17     208 non-null    float64
 17  V18     208 non-null    float64
 18  V19     208 non-null    float64
 19  V20     208 non-null    float64
 20  V21     208 non-null    float64
 21  V22     208 non-null    float64
 22  V2

In [8]:
print(dataset['Class'].value_counts())

0    111
1     97
Name: Class, dtype: int64


## Correlation of the features on dataset

In [9]:
correlation_matrix = dataset.corr()
corr = correlation_matrix['Class'].sort_values(ascending=False)
correlation_dataframe = pd.DataFrame({'column': corr.index,
                 'Correlation with target': corr.values})
correlation_dataframe

column  Correlation with target
0   Class                 1.000000
1     V36                 0.269151
2     V35                 0.227670
3     V37                 0.209055
4     V34                 0.172010
..    ...                      ...
56    V45                -0.339406
57    V10                -0.341142
58    V49                -0.351312
59    V12                -0.392245
60    V11                -0.432855

[61 rows x 2 columns]

In [7]:
dataset.describe()

V1          V2          V3          V4          V5          V6  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               V7          V8          V9         V10  ...         V52  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.013420   
std      0.061788    0.085152    0.118387    0.134416  ...    0.009634   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000800   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.007275   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.011400   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.016725   
max      0.372900    0.459000    0.682800    0.710600  ...    0.070900   

              V53         V54         V55         V56         V57         V58  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.010709    0.010941    0.009290    0.008222    0.007820    0.007949   
std      0.007060    0.007301    0.007088    0.005736    0.005785    0.006470   
min      0.000500    0.001000    0.000600    0.000400    0.000300    0.000300   
25%      0.005075    0.005375    0.004150    0.004400    0.003700    0.003600   
50%      0.009550    0.009300    0.007500    0.006850    0.005950    0.005800   
75%      0.014900    0.014500    0.012100    0.010575    0.010425    0.010350   
max      0.039000    0.035200    0.044700    0.039400    0.035500    0.044000   

              V59         V60       Class  
count  208.000000  208.000000  208.000000  
mean     0.007941    0.006507    0.466346  
std      0.006181    0.005031    0.500070  
min      0.000100    0.000600    0.000000  
25%      0.003675    0.003100    0.000000  
50%      0.006400    0.005300    0.000000  
75%      0.010325    0.008525    1.000000  
max      0.036400    0.043900    1.000000  

[8 rows x 61 columns]

# Feature Engineering

In [10]:
label = dataset['Class'] #--> Target Prediksi
feature_used = dataset.drop(['Class',],axis=1) #--> Fitur yang digunakan adalah selain fitur 'target'

# Menggunakan 2 metode scalling, Standard dan Power Transform dengan yeo-jhonson
scaler_1 = StandardScaler(with_std=True,with_mean=True)
scaler_2 = PowerTransformer(method='yeo-johnson', standardize=True)

x_1 = scaler_1.fit_transform(feature_used)
x_2 = scaler_2.fit_transform(feature_used)

data_x_1 = pd.DataFrame(x_1, columns=feature_used.columns)
data_x_2 = pd.DataFrame(x_2, columns=feature_used.columns)

In [11]:
def scores(obj, Predict, Feature2, Label2):
    print('Accuracy   on test set: {:.3f}'.format(obj.score(Feature2, Label2)))
    print('F1_score   on test set: {:.3f}'.format(f1_score(Label2, Predict, average='macro')))
    print('Precision  on test set: {:.3f}'.format(precision_score(Label2, Predict, average='macro')))
    print('Recall     on test set: {:.3f}'.format(recall_score(Label2, Predict, average='macro')))

# Split Data 80-20

# Logistic Regression

In [13]:
# Training menggunakan data hasil scalling dengan Standard Scaler
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)

# Logistic Regression
print("Menggunakan Standard Scaler")
logReg = LogisticRegression()
logReg.fit(feature_train,label_train)
cross_val= cross_val_score(logReg, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= logReg.predict(feature_test)
scores(logReg, prediction, feature_test, label_test)

print("-----------------------------------------------------------------------------------------------------------------------------")
print()

print("Menggunakan PowerTransform Scaller")
# Training menggunakan data hasil scalling dengan Power Transform Scaler
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20,random_state=4)
logReg = LogisticRegression()
logReg.fit(feature_train,label_train)
cross_val_= cross_val_score(logReg, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= logReg.predict(feature_test)
scores(logReg, prediction, feature_test, label_test)


Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.810
F1_score   on test set: 0.802
Precision  on test set: 0.802
Recall     on test set: 0.802
-----------------------------------------------------------------------------------------------------------------------------

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.810
F1_score   on test set: 0.802
Precision  on test set: 0.802
Recall     on test set: 0.802


# MLP Classifier

In [16]:
print("Menggunakan Standard Scaler")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
mlc = MLPClassifier(hidden_layer_sizes=100, activation='relu',solver='lbfgs',batch_size='auto', learning_rate_init=0.0001, max_iter=10000,early_stopping=False)
mlc.fit(feature_train,label_train)
cross_val_= cross_val_score(mlc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= mlc.predict(feature_test)
scores(mlc, prediction, feature_test, label_test)

print("-----------------------------------------------------------------------------------------------------------------------------")
print()

print("Menggunakan PowerTransform Scaller")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
mlc= MLPClassifier(hidden_layer_sizes=100, activation='relu',solver='lbfgs',batch_size='auto', learning_rate_init=0.0001, max_iter=10000, early_stopping=False)
mlc.fit(feature_train,label_train)
cross_val_= cross_val_score(mlc, feature_train, label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= mlc.predict(feature_test)
scores(mlc, prediction, feature_test, label_test)

Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.905
F1_score   on test set: 0.896
Precision  on test set: 0.931
Recall     on test set: 0.882
-----------------------------------------------------------------------------------------------------------------------------

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.905
F1_score   on test set: 0.899
Precision  on test set: 0.911
Recall     on test set: 0.892


# Random Forest Classifier  

In [17]:
estimator =[5,10,20,30,50,75,100,150]
for i in estimator:
    print("Menggunakan Standard Scaler")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
    print("Untuk Estimator "+str(i))
    rfc = RandomForestClassifier(i,criterion='gini',bootstrap=True,class_weight='balanced_subsample')
    rfc.fit(feature_train,label_train)
    cross_val_= cross_val_score(rfc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= rfc.predict(feature_test)
    scores(rfc, prediction, feature_test, label_test)
    print()

    print("Menggunakan PowerTransform Scaller")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
    rfc = RandomForestClassifier(i,criterion='gini',bootstrap=True,class_weight='balanced_subsample')
    rfc.fit(feature_train,label_train)
    cross_val_= cross_val_score(rfc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= rfc.predict(feature_test)
    scores(rfc, prediction, feature_test, label_test)
    print()
    print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan Standard Scaler
Untuk Estimator 5
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.833
F1_score   on test set: 0.825
Precision  on test set: 0.829
Recall     on test set: 0.822

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.690
F1_score   on test set: 0.658
Precision  on test set: 0.683
Recall     on test set: 0.655

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan Standard Scaler
Untuk Estimator 10
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.810
F1_score   on test set: 0.802
Precision  on test set: 0.802
Recall     on test set: 0.802

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test s

# Gaussian-Process Classifier 

In [18]:
print("Menggunakan Standard Scaler")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
kernel = 1.0 * RBF(1.0)
gpc = GaussianProcessClassifier(kernel=kernel, max_iter_predict=100, optimizer='fmin_l_bfgs_b')
gpc.fit(feature_train,label_train)
cross_val_= cross_val_score(gpc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gpc.predict(feature_test)
scores(gpc, prediction, feature_test, label_test)
print()

print("Menggunakan Power Transform Scaler")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
kernel = 1.0 * RBF(1.0)
gpc = GaussianProcessClassifier(kernel=kernel, max_iter_predict=100,optimizer='fmin_l_bfgs_b')
gpc.fit(feature_train,label_train)
cross_val_= cross_val_score(gpc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gpc.predict(feature_test)
scores(gpc, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.905
F1_score   on test set: 0.896
Precision  on test set: 0.931
Recall     on test set: 0.882

Menggunakan Power Transform Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.857
F1_score   on test set: 0.844
Precision  on test set: 0.875
Recall     on test set: 0.833



# KNN Classifier 

In [19]:
numbers = 16
for i in range(numbers):
    if i %2 != 0 :
        print("Menggunakan nilai k : "+str(i))
        print("Menggunakan Standard Scaler")
        feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
        knn = KNeighborsClassifier(i,weights='distance',algorithm='auto')
        knn.fit(feature_train,label_train)
        cross_val_= cross_val_score(knn, feature_train,label_train, cv=5)
        print("Cross Validation Score : "+str(cross_val))
        prediction= knn.predict(feature_test)
        scores(knn, prediction, feature_test, label_test)
        print()

        print("Menggunakan Power Transform Scaler")
        feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
        knn = KNeighborsClassifier(i,weights='distance',algorithm='auto')
        knn.fit(feature_train,label_train)
        cross_val_= cross_val_score(knn, feature_train,label_train, cv=5)
        print("Cross Validation Score : "+str(cross_val))
        prediction= knn.predict(feature_test)
        scores(knn, prediction, feature_test, label_test)
        print()
        print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan nilai k : 1
Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.786
F1_score   on test set: 0.770
Precision  on test set: 0.786
Recall     on test set: 0.764

Menggunakan Power Transform Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.833
F1_score   on test set: 0.821
Precision  on test set: 0.839
Recall     on test set: 0.813

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan nilai k : 3
Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.786
F1_score   on test set: 0.763
Precision  on test set: 0.800
Recall     on test set: 0.754

Menggunakan Power Transform Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy 

# Adaboost Classifier

In [22]:
estimator =[5,10,20,30,50,75,100,150]
for i in estimator:
    print("Menggunakan Standard Scaler")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
    print("Untuk Estimator "+str(i))
    adc = AdaBoostClassifier(n_estimators=i,learning_rate=0.001, random_state=100,algorithm='SAMME.R')
    adc.fit(feature_train,label_train)
    cross_val_= cross_val_score(adc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= adc.predict(feature_test)
    scores(adc, prediction, feature_test, label_test)
    print()

    print("Menggunakan PowerTransform Scaller")
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
    adc = AdaBoostClassifier(n_estimators=i,learning_rate=0.001, random_state=100,algorithm='SAMME.R')
    adc.fit(feature_train,label_train)
    cross_val_= cross_val_score(adc, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= adc.predict(feature_test)
    scores(adc, prediction, feature_test, label_test)
    print()
    print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan Standard Scaler
Untuk Estimator 5
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.762
F1_score   on test set: 0.721
Precision  on test set: 0.808
Recall     on test set: 0.715

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.762
F1_score   on test set: 0.721
Precision  on test set: 0.808
Recall     on test set: 0.715

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan Standard Scaler
Untuk Estimator 10
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.762
F1_score   on test set: 0.721
Precision  on test set: 0.808
Recall     on test set: 0.715

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test s

# Decision Tree Classifier

In [23]:
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
print("Menggunakan Standard Scaler")
dtc=DecisionTreeClassifier(criterion='entropy',splitter='best',class_weight='balanced')
dtc.fit(feature_train,label_train)
cross_val_= cross_val_score(dtc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= dtc.predict(feature_test)
scores(dtc, prediction, feature_test, label_test)
print()

print("Menggunakan PowerTransform Scaller")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
dtc = DecisionTreeClassifier(criterion='entropy',splitter='best',class_weight='balanced')
dtc.fit(feature_train,label_train)
cross_val_= cross_val_score(dtc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= dtc.predict(feature_test)
scores(dtc, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.714
F1_score   on test set: 0.704
Precision  on test set: 0.704
Recall     on test set: 0.704

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.762
F1_score   on test set: 0.748
Precision  on test set: 0.756
Recall     on test set: 0.744



# Gaussian Naive-Bayes

In [31]:
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
print("Menggunakan Standard Scaler")
gnb=GaussianNB()
gnb.fit(feature_train,label_train)
cross_val_= cross_val_score(gnb, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gnb.predict(feature_test)
scores(gnb, prediction, feature_test, label_test)
print()

print("Menggunakan PowerTransform Scaller")
feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
gnb=GaussianNB()
gnb.fit(feature_train,label_train)
cross_val_= cross_val_score(gnb, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= gnb.predict(feature_test)
scores(gnb, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.619
F1_score   on test set: 0.618
Precision  on test set: 0.659
Recall     on test set: 0.652

Menggunakan PowerTransform Scaller
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.643
F1_score   on test set: 0.633
Precision  on test set: 0.632
Recall     on test set: 0.634



# SVC

In [32]:
kernels=['poly','linear','rbf']
for i, value in enumerate(kernels):
    print("Menggunakan kernel : "+str(value))
    feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
    print("Menggunakan Standard Scaler")
    svl=SVC(kernel= value, C=1000,gamma='scale',tol=0.0001,class_weight='balanced',degree=3,coef0=1)
    svl.fit(feature_train,label_train)
    cross_val_= cross_val_score(svl, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= svl.predict(feature_test)
    scores(svl, prediction, feature_test, label_test)
    print()

    feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
    print("Menggunakan Power Transform Scaler")
    svl=SVC(kernel= value, C=1000,gamma='scale',tol=0.0001,class_weight='balanced',degree=3,coef0=1)
    svl.fit(feature_train,label_train)
    cross_val_= cross_val_score(svl, feature_train,label_train, cv=5)
    print("Cross Validation Score : "+str(cross_val))
    prediction= svl.predict(feature_test)
    scores(svl, prediction, feature_test, label_test)
    print()
    print("-----------------------------------------------------------------------------------------------------------------------------")

Menggunakan kernel : poly
Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.905
F1_score   on test set: 0.899
Precision  on test set: 0.911
Recall     on test set: 0.892

Menggunakan Power Transform Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.905
F1_score   on test set: 0.896
Precision  on test set: 0.931
Recall     on test set: 0.882

-----------------------------------------------------------------------------------------------------------------------------
Menggunakan kernel : linear
Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.762
F1_score   on test set: 0.757
Precision  on test set: 0.755
Recall     on test set: 0.762

Menggunakan Power Transform Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Acc

# LinearSVC

In [33]:
feature_train, feature_test, label_train, label_test = train_test_split(data_x_1, label, test_size = 0.20, random_state=4)
print("Menggunakan Standard Scaler")
linearsvc = LinearSVC(class_weight='balanced',max_iter=1000)
linearsvc.fit(feature_train,label_train)
cross_val_= cross_val_score(linearsvc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= linearsvc.predict(feature_test)
scores(svl, prediction, feature_test, label_test)
print()

feature_train, feature_test, label_train, label_test = train_test_split(data_x_2, label, test_size = 0.20, random_state=4)
print("Menggunakan Power Transform Scaler")
linearsvc = LinearSVC(class_weight='balanced',max_iter=1000)
linearsvc.fit(feature_train,label_train)
cross_val_= cross_val_score(linearsvc, feature_train,label_train, cv=5)
print("Cross Validation Score : "+str(cross_val))
prediction= linearsvc.predict(feature_test)
scores(svl, prediction, feature_test, label_test)
print()

Menggunakan Standard Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.929
F1_score   on test set: 0.802
Precision  on test set: 0.802
Recall     on test set: 0.802

Menggunakan Power Transform Scaler
Cross Validation Score : [0.70588235 0.57575758 0.75757576 0.78787879 0.78787879]
Accuracy   on test set: 0.905
F1_score   on test set: 0.757
Precision  on test set: 0.755
Recall     on test set: 0.762



# Kesimpulan

Pada percobaan diatas, menggunakan dataset 'Diabetes'. digunakan 10 algoritma untuk mengklasifikasi data menjadi 2 kelas
yaitu 0 = 'Tidak terkena Diabetes' , 1 = 'Terkena Diabetes' .

Dataset tersebut di-scaling dengan 2 scaler, yaitu 'Standard scaler' dan 'Power Transform'.

* Algoritma LogisticRegression mendapat nilai akurasi dan precision

    Standard scaler : 0.810 dan 0.802 , Power Transform : 0.810 dan 0,802

* Algoritma MLP Classifier mendapat nilai akurasi dan precision

    Standard scaler : 0.905 dan 0.931 , Power Transform : 0.905 dan 0.911

* Algoritma RandomForest Classifier mendapat nilai akurasi dan precision
yang terbaik ada pada nilai estimator 100, dengan :

    Standard scaler : 0.881 dan 0.893 , Power Transform : 0.881 dan 0.893

* Algoritma KNN Classifier mendapat nilai akurasi dan precision
yang terbaik ada pada nilai k = 99, dengan :

    Standard scaler : 0.857 dan 0.875 , Power Transform : 0.857 dan 0.875

* Algoritma Adaboost Classifier mendapat nilai akurasi dan precision
yang terbaik ada pada nilai estimator 5, dengan :

    Standard scaler : 0.762 dan 0.808 , Power Transform : 0.762 dan 0.808

* Algoritma Decision Tree Classifier mendapat nilai akurasi dan precision

    Standard scaler : 0.714 dan 0.704 , Power Transform : 0.762 dan 0.756

* Algoritma Gaussian Naive-Bayes mendapat nilai akurasi dan precision

    Standard scaler : 0.619 dan 0.659 , Power Transform : 0.643 dan 0.632

* Algoritma SVC yang mendapat nilai terbaik menggunakan kernel 'rbf'
dengan nilai akurasi dan precision

    Standard scaler : 0.929 dan 0.946 , Power Transform : 0.905 dan 0.931

* Algoritma LinearSVC mendapat nilai akurasi dan precision

    Standard scaler : 0.929 dan 0.802 , Power Transform : 0.905, 0.755

Dari hasil percobaan diatas, algoritma SVC-rbf dan LinearSVC, menggunakan data yang
di-scaling dengan Standard scaler mendapat hasil akurasi yang terbaik sebesar 0.929

Maka, dapat disimpulkan bahwa penggunaan Standard scaler pada dataset ini sudah mencukupi
untuk bisa mendapat nilai akurasi dan precision yang cukup baik pada kebanyakan algoritma yang digunakan. 
Tapi, dalam beberapa algoritma akan lebih baik hasilnya jika menggunakan Power Transform scaler.

Hasil klasifikasi yang didapatkan oleh model algoritma dapat dipengaruhi oleh banyak faktor,
salah satunya adalah penggunaan Hyperparameter pada tiap algoritma dan keseimbangan jumlah sebaran data pada label.
jika seimbang, maka pengklasifikasian data akan terjadi merata / tidak condong pada 1 kelas.